### DISCUSSION TITLE
Slow serial communication with Bonsai

### DISCUSSION THREAD

**Discussion by liadJB** [bonsai-rx/bonsai#984] [Q&A]
Hi,

 I am using the following script to communicate with Arduino (Mega 2560)
[ArduinoExample.zip](https://github.com/bonsai-rx/bonsai/files/9188742/ArduinoExample.zip)
Looking at the timestamps, there are two problems. 
The first, it seems that the acquisition rate is not stable. That is, the difference between the timestamps fluctuates quite widely between 0.5 ms to 4 ms. 
The fluctuation is surmountable. However, the mean difference between timestamps is roughly equivalent to 250 Hz. 
Ideally I would want to acquire data at a faster rate (~1000 Hz would be ideal).
I believe the latency comes from Bonsai rather than from Arduino, since lowering the baudrate decreases the sampling rate, but no matter how high the baud rate is it does not go past 250 Hz. 
Is there any way to configure the communication in Bonsai to make it more efficient?

Thanks
Liad

**Comment by bruno-f-cruz** [comment]
Hey, can you attach the code running in the Arduino? thanks

**Reply by liadJB** [reply]
Yep:

See below:



#define wheelPin_1 5        // Wheel
#define wheelPin_2 3        // Wheel
#define cameraPin 7             // digital pin of sync
#define syncPin 3             // digital pin of sync

/*
int photodiodePin = 0;    
int syncPin = 0;
*/

unsigned int wheelVal1;
unsigned int wheelVal2;
unsigned int CameraVal;

unsigned long lastTime = millis();
unsigned int toggle = 0 ;

void setup()

{
  pinMode(wheelPin_1, INPUT);   // PhotoDiode
  pinMode(wheelPin_1, INPUT);   // digital pin of sync
  pinMode(cameraPin, INPUT);
  pinMode(syncPin,OUTPUT);

  digitalWrite(syncPin,LOW);
  
  Serial.begin(115200);          //  setup serial
  delay(500);
}



void loop()

{

  wheelVal1 = analogRead(wheelPin_1);    // read the input pin
  wheelVal2 = analogRead(wheelPin_2);
  CameraVal = analogRead(cameraPin);

  if ((millis()-lastTime)>random(5,20))
  {
    toggle = !toggle;
    digitalWrite(syncPin,toggle);
    lastTime  = millis();
  }

  Serial.print(wheelVal1);
  Serial.print("\t");
  Serial.print(wheelVal2);
  Serial.print("\t");
  Serial.print(CameraVal);
  Serial.print("\t");
  Serial.print(CameraVa2);
  Serial.print("\t");
  Serial.println(toggle);

  delay(1);
}

**Reply by bruno-f-cruz** [reply]
The true acquisition frequency is determined in the Arduino not in Bonsai. When you timestamp in bonsai you are timestamping when the sample arrived to the software not when it was collected. If you want to have a better sense of time between samples add a:

"Serial.println(millis());" to your aduino code and parse it in bonsai. That should give you a better estimate of sampling rate.

Relative to getting faster sampling rates, that's something that, for the most part, you will have to optimize in your Arduino code.  Disabling the "delay(1)" should get you faster sampling for instance. Otherwise you could look at other optimization strategies (e.g. https://forum.arduino.cc/t/reading-2-analog-inputs-at-1khz-interrupts-timed-loop/887469).

Good luck!

**Comment by liadJB** [comment]
Hi Bruno,

Thanks for the reply. I now changed how acquire the time, and indeed it makes more sense (more stable ~500Hz acquisition).
